File: backend/models/personnel.go
package models

import "time"

// PersonnelStatus บอกว่าเจ้าหน้าที่คนนี้ยังปฏิบัติงานอยู่หรือไม่
type PersonnelStatus string

const (
	PersonnelActive   PersonnelStatus = "active"
	PersonnelInactive PersonnelStatus = "inactive"
)

// IsValid กันค่าสถานะแปลกปลอมที่ส่งมาจากหน้าเว็บ
func (s PersonnelStatus) IsValid() bool {
	return s == PersonnelActive || s == PersonnelInactive
}

// Personnel คือเจ้าหน้าที่หนึ่งคนในหอสมุด แยกจากตาราง users ที่ใช้ล็อกอิน
// คนที่มีแถวที่นี่อาจยังไม่มีบัญชีล็อกอินก็ได้ และกลับกัน
type Personnel struct {
	ID         int    `gorm:"primaryKey" json:"id"`
	StaffID    string `gorm:"size:30;uniqueIndex;not null" json:"staff_id"`
	FirstName  string `gorm:"size:100;not null" json:"first_name"`
	LastName   string `gorm:"size:100;not null" json:"last_name"`
	Department string `gorm:"size:120;index" json:"department"`
	Position   string `gorm:"size:120" json:"position"`
	Email      string `gorm:"size:150" json:"email"`
	Phone      string `gorm:"size:30" json:"phone"`

	// StartDate เก็บเป็นข้อความรูปแบบ YYYY-MM-DD ไม่ใช่ timestamp
	// เพราะเป็นวันที่ล้วนไม่มีเวลา ถ้าเก็บเป็น timestamp จะเพี้ยนไปหนึ่งวันเวลาข้ามโซนเวลา
	StartDate string `gorm:"size:10" json:"start_date"`

	Status PersonnelStatus `gorm:"size:20;index" json:"status"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

File: frontend/src/services/https/personnel.ts
import { apiFetch } from './index'
import type { Personnel, PersonnelFormData, PersonnelStatus } from '../../interface/IPersonnelInterface'

interface PersonnelResponse {
  id: number
  staff_id: string
  first_name: string
  last_name: string
  department: string
  position: string
  email: string
  phone: string
  start_date: string
  status: PersonnelStatus
}

export function toPersonnel(res: PersonnelResponse): Personnel {
  return {
    id: res.id,
    staffId: res.staff_id,
    firstName: res.first_name,
    lastName: res.last_name,
    department: res.department,
    position: res.position,
    email: res.email,
    phone: res.phone,
    startDate: res.start_date,
    status: res.status,
  }
}

function toRequest(data: PersonnelFormData) {
  return {
    staff_id: data.staffId,
    first_name: data.firstName,
    last_name: data.lastName,
    department: data.department,
    position: data.position,
    email: data.email,
    phone: data.phone,
    start_date: data.startDate,
    status: data.status,
  }
}

export async function listPersonnel(token: string): Promise<Personnel[]> {
  const people = await apiFetch<PersonnelResponse[]>('/api/v1/personnel', { token })
  return people.map(toPersonnel)
}

export async function createPersonnel(token: string, data: PersonnelFormData): Promise<Personnel> {
  const created = await apiFetch<PersonnelResponse>('/api/v1/personnel', {
    method: 'POST',
    token,
    body: toRequest(data),
  })
  return toPersonnel(created)
}

export async function updatePersonnel(
  token: string,
  id: number,
  data: PersonnelFormData,
): Promise<Personnel> {
  const updated = await apiFetch<PersonnelResponse>(`/api/v1/personnel/${id}`, {
    method: 'PATCH',
    token,
    body: toRequest(data),
  })
  return toPersonnel(updated)
}

export function deletePersonnel(token: string, id: number) {
  return apiFetch<{ message: string }>(`/api/v1/personnel/${id}`, { method: 'DELETE', token })
}

export async function togglePersonnelStatus(token: string, id: number): Promise<Personnel> {
  const person = await apiFetch<PersonnelResponse>(`/api/v1/personnel/${id}/status`, {
    method: 'POST',
    token,
  })
  return toPersonnel(person)
}
